src/backend/mpi/swdsm.cpp

```diff
@@ -419,85 +412,10 @@ void handler(int sig, siginfo_t *si, void *unused){
 	tag = cacheControl[startIndex].tag;
 
 	if(state == INVALID || (tag != lineAddr && tag != GLOBAL_NULL)){
-		if(loadline[0] == GLOBAL_NULL && prefetchline[0] != GLOBAL_NULL){
-			if(prefetchline[0] <= startIndex && startIndex < prefetchline[0]+CACHELINE){
-				loadline[0]= (startIndex+CACHELINE)%cachesize;
-				loadtag[0]=lineAddr+pagesize*CACHELINE;
-			}
-			else{
-				loadline[0]=cacheIndex%cachesize;
-				loadtag[0]=alignedDistrAddr;
-			}
-
-			sem_post(&loadstartsem);
-			sem_wait(&prefetchwaitsem);
-			prefetchline[0] = GLOBAL_NULL;
-			prefetchtag[0]= GLOBAL_NULL;
-			pthread_mutex_unlock(&cachemutex);
-			double t2 = MPI_Wtime();
-			stats.loadtime+=t2-t1;
-			return;
-		}
-		else if(prefetchline[0] == GLOBAL_NULL && loadline[0] != GLOBAL_NULL){
-			if(loadline[0] <= startIndex && startIndex < loadline[0]+CACHELINE){
-				prefetchline[0]=(startIndex+CACHELINE)%cachesize;
-				prefetchtag[0]=lineAddr+pagesize*CACHELINE;
-			}
-			else{
-				prefetchline[0]=cacheIndex%cachesize;
-				prefetchtag[0]=alignedDistrAddr;
-			}
-
-			sem_post(&prefetchstartsem);
-			sem_wait(&loadwaitsem);
-			loadline[0] = GLOBAL_NULL;
-			loadtag[0]=GLOBAL_NULL;
-			pthread_mutex_unlock(&cachemutex);
-			double t2 = MPI_Wtime();
-			stats.loadtime+=t2-t1;
-			return;
-		}
-		else{
-			loadline[0]=startIndex%cachesize;
-			loadtag[0]=alignedDistrAddr;
-			sem_post(&loadstartsem);
-
+		load_cache_entry(alignedDistrAddr, (startIndex%cachesize));
 #ifdef DUAL_LOAD
-			prefetchline[0]=(startIndex+CACHELINE)%cachesize;
-			prefetchtag[0]=alignedDistrAddr+CACHELINE*pagesize;
-			sem_post(&prefetchstartsem);
-			sem_wait(&loadwaitsem);
-			loadline[0]=GLOBAL_NULL;
-			loadtag[0]=GLOBAL_NULL;
-#else
-			sem_wait(&loadwaitsem);
-			loadline[0]=GLOBAL_NULL;
-			loadtag[0]=GLOBAL_NULL;
-			prefetchline[0]=GLOBAL_NULL;
-			prefetchtag[0]=GLOBAL_NULL;
-
+		prefetch_cache_entry((alignedDistrAddr+CACHELINE*pagesize), ((startIndex+CACHELINE)%cachesize));
```

Comment:
Would `(startIndex%cachesize) + 1` be more understandable? This is a minor thing, not really a blocking issue.

Reply:
Given that CACHELINE is defined as 1L, your suggestion would allow the entry to go out of bounds, no?

Reply:
Right, my bad